xlators/storage/posix/src/posix-common.c

```diff
@@ -1231,6 +1266,111 @@ posix_init(xlator_t *this)
                    out);
 
     GF_OPTION_INIT("ctime", _private->ctime, bool, out);
+    /*Check the trusted.graceful_reboot xattr */
+    ret = sys_lgetxattr(_private->base_path, "trusted.graceful_reboot",
+                        xattr_value, 1);
+    if (xattr_value[0] == '0') {
+        gf_log(this->name, GF_LOG_INFO, "Last brick reboot was not graceful");
+        for (i = 0; i < 256; i++) {
+            snprintf(dir_name, sizeof(dir_name), "%s/%02x", dir_handle, i);
+            drptr = sys_opendir(dir_name);
+            while ((dentry = sys_readdir(drptr, scratch)) != NULL) {
+                dentry_len = strlen(dentry->d_name);
+                if (dentry_len < UUID_CANONICAL_FORM_LEN)
+                    continue;
+
+                /* if dentry_name exists rename as a suffix it means rename dir
+                   fop was not succeeded completly so need to validate gfid and
+                   heal if it was not linked properly
+                */
+                if (dentry_len == 43) {
+                    /* Need to follow below steps to validate the same
+                       1) Populate rename_gfid_str after ignore _rename suffix
+                       2) Generate rename_gfid from rename_gfid_str
+                       3) Populate rename_gfid_path and gfid_path that was
+                       created by posix_rename before calling sys_rename on
+                       original path 4) Read link of rename_gfid_path to
+                       validate gfid link 5) Fetch gfid from link path and
+                       compare with rename_gfid 6) If gfid's are not matching it
+                       means the crash was happend before call rename for
+                       original path means rename was not executed need to
+                       delete dentry name and rename_gfid_path 7) If gfid's are
+                       matching it means crash was happened after call rename
+                       for original path before rename gfid path so need to heal
+                       gfid
+                    */
+                    strncpy(rename_gfid_str, dentry->d_name,
+                            UUID_CANONICAL_FORM_LEN);
+                    gf_uuid_parse(rename_gfid_str, rename_gfid);
+                    snprintf(rename_gfid_path, sizeof(rename_gfid_path),
+                             "%s/%02x/%02x/%s_rename", dir_handle,
+                             rename_gfid[0], rename_gfid[1], rename_gfid_str);
+                    snprintf(gfid_path, sizeof(gfid_path), "%s/%02x/%02x/%s",
```

Comment:
nit: I think you can defer this snprintf() to later in the code, after line 1326?